### PR TITLE
util: use stronger-guarantee rename method

### PR DESCRIPTION
**Pull request by vasild**
Use `std::filesystem::rename()` instead of `std::rename()`. We rely on the
destination to be overwritten if it exists, but `std::rename()`'s behavior
is implementation-defined in this case.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Please update doc/dependencies to gcc 8 (from 7), as std::fs is not included in gcc 7

**Comment by vasild** [comment]
Updated. I am not sure if this PR shouldn't be put on hold until CentOS 23 starts shipping with a recent enough compiler?

**Comment by MarcoFalke** [comment]
Maybe adding `-lstdc++fs` to the build flags via configure by default fixes the compile errors?

**Comment by laanwj** [comment]
Concept ACK! That's another WIN32 specific hack less.

But yes it's probably going to be a looong time before this can be merged.

**Comment by vasild** [comment]
I will leave this as is (open, with failing CI) until the compilers we support catch up with C++17 filesystem support.

**Comment by vasild** [comment]
Removed the changes to `doc/dependencies.md` as those will be carried by https://github.com/bitcoin/bitcoin/issues/20460 (bumping minimum compiler version requirements), also they caused conflicts.

**Comment by luke-jr** [comment]
:>Removed the changes to doc/dependencies.md as those will be carried by #20460 (bumping minimum compiler version requirements), also they caused conflicts.

You should probably rebase this PR on top of that then...

**Comment by vasild** [comment]
@luke-jr that's right, but #20460 is just an issue/idea, without a patch (yet)...

**Comment by laanwj** [comment]
> I will leave this as is (open, with failing CI) until the compilers we support catch up with C++17 filesystem support.

I don't labeling this as "waiting for author" is quite right.

**Comment by MarcoFalke** [comment]
This pull is waiting for the author to rebase on the pull that fixes #20460  . Is there a better label?

**Comment by vasild** [comment]
There is no fix for the issue #20460 yet.

This PR is waiting for newer compilers to hit the `master` branch (docs and ci).

**Comment by fanquake** [comment]
> There is no fix for the issue #20460 yet.

If you want, you can now rebase on top of #20744.

File: src/util/system.cpp
Comment (MarcoFalke) [inline src/util/system.cpp:1013]: Shouldn't doxygen be in the `.h` file?

**Comment by MarcoFalke** [comment]
> If you want, you can now rebase on top of #20744.

I think a rebase is not (strictly) needed, just a force push to re-trigger CI.

**Comment by fanquake** [comment]
> I think a rebase is not (strictly) needed, just a (force) push to re-trigger CI.

Sure, but this can't be merged before #20744 in any case.

**Comment by vasild** [comment]
`166b1542d7...99b151df9c`: rebase

> This PR is waiting for newer compilers to hit the `master` branch

https://github.com/bitcoin/bitcoin/pull/23060 did just that (merged).

I don't think this PR depends on https://github.com/bitcoin/bitcoin/pull/20744. @fanquake, why do you think this PR cannot be merged before https://github.com/bitcoin/bitcoin/pull/20744?

File: src/util/system.cpp
Comment (vasild) [reply src/util/system.cpp]: Right, moved.

**Comment by fanquake** [comment]
> I don't think this PR depends on #20744. @fanquake, why do you think this PR cannot be merged before #20744?

Because we are either using `boost::filesystem` or `std::filesystem`. We're not going to have the repo exist in some awkward middle ground where we are using both. In any case, using `std::filesystem` is not necessarily straight forward, and likely requires the changes in #22937 before it can be used.

**Comment by fanquake** [comment]
Another reason is that this doesn't have any build system support for properly using `std::filesystem`, and the additional linker flags that may be needed, i.e https://github.com/bitcoin/bitcoin/pull/20744/commits/721928821f1623265db9c07f2fcef49a356d477f.

File: src/util/system.cpp
Comment (ryanofsky) [inline src/util/system.cpp:1069]: In commit "util: use stronger-guarantee rename method" (99b151df9c074f60150c5eb777ed98bfd7118e91)

This will result in narrow string conversion on windows, and after #22937 will be a compiler error. Should be easy to fix by changing to:

```c++
std::filesystem::rename(src.native(), dest.native(), error);
```

This should also work after #20744, but after #20744 it could be simplified more to

```c++
std::filesystem::rename(src, dest, error);
```

**Review by ryanofsky — APPROVED** [review]
Conditional code review ACK 99b151df9c074f60150c5eb777ed98bfd7118e91 if this is rebased and `.string()` is replaced by `.native()` (see comment below).

This is just a review of the c++ code, disregarding build issues. IIUC the new rename function may not be callable on all platforms without some of the build changes in #20744.

File: src/util/system.cpp
Comment (vasild) [reply src/util/system.cpp]: Used just `src` and `dest`, thanks!

**Review by ryanofsky — APPROVED** [review]
Code review ACK a9ab42ca973e1841ac125b0f64eaa922df240c08 assuming #20744 is merged first (it requires #20744 to build)

**Comment by fanquake** [comment]
I've rebased this in #24308.

**Comment by fanquake** [comment]
Going to close this in favour of #24308.

**Pull request closed without merging**
